util/include/Serializable.h

```diff
@@ -37,72 +40,199 @@ typedef std::unique_ptr<char[], std::default_delete<char[]>> UniquePtrToChar;
 typedef std::unique_ptr<unsigned char[], std::default_delete<unsigned char[]>> UniquePtrToUChar;
 typedef std::shared_ptr<Serializable> SerializablePtr;
 
-//TODO [TK] remove
-class MemoryBasedBuf : public std::basic_streambuf<char> {
- public:
-  MemoryBasedBuf(const UniquePtrToChar &buf, size_t size) {
-    setg(buf.get(), buf.get(), buf.get() + size);
-  }
-};
-
-// This class allows usage of a regular buffer as a stream.
-// TODO [TK] remove
-class MemoryBasedStream : public std::istream {
- public:
-  MemoryBasedStream(const UniquePtrToChar &buf, size_t size) : std::istream(&buffer_), buffer_(buf, size) {
-    rdbuf(&buffer_);
-  }
-
- private:
-  MemoryBasedBuf buffer_;
-};
-
 class Serializable {
- public:
-
-  Serializable();
+public:
+  Serializable() = default;
   virtual ~Serializable() = default;
+  /**
+   * serialization API
+   */
+  template<typename T>
+  static void serialize(std::ostream& outStream, const T& t) { serialize_impl(outStream, t, int{}); }
+  template<typename T>
+  /**
+   * deserialization API
+   */
+  static void deserialize(std::istream& inStream, T& t) { deserialize_impl(inStream, t, int{}); }
+  /**
+   * convenience serialization function
+   */
+  virtual void serialize(std::ostream& outStream) const final{ serialize(outStream, *this); }
+
+protected:
+  /**
+   * the class version
+   */
+  virtual const std::string getVersion()  const = 0;
+  /**
+   * each class knows how to serialize its data members
+   */
+  virtual void serializeDataMembers  (std::ostream&) const = 0;
+  /**
+   * each class knows how to deserialize its data members
+   */
+  virtual void deserializeDataMembers(std::istream&)       = 0;
+
+  virtual const std::string getName() const final {return typeid(*this).name();}
+
+protected:
+  /** *****************************************************************************************************************
+   *  std::container
+   *  - container size
+   *  - every element of the container (recursively)
+   */
+  template<typename T, typename std::enable_if<is_std_container<T>::value, T>::type* = nullptr>
+  static void serialize_impl(std::ostream& outStream, const T& container, int) {
+   typename T::size_type size = container.size();
```

Comment:
When we serialize sizes, should we try to make them platform independent, by serializing as a uint64_t rather than a size_t ?

Reply:
it's a size_type of a container.

Reply:
I get that. But if this data is ever transferred to a machine with a different architecture, the size_type will be different, and it may be read back incorrectly.